### PR TITLE
initial thoughts on code to ensure reserved values exist

### PR DESCRIPTION
**Pull request by eileenmcnaughton**

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@totten - I need to create an activity type option value - but have the risk it could already exist. I feel like we could do with a function that gets called on upgrade & checks reserved option values exist as we've hit this before. This PR is more a starting point to start thinking about it - as it puts the metadata directly into the function -

File: CRM/Core/BAO/OptionValue.php [hunk 1]
```diff
@@ -518,4 +518,38 @@ public static function getOptionValuesAssocArrayFromName($optionGroupName) {
     return $options;
   }
 
+  /**
+   * Ensure that declared reserved option values exist.
+   */
+  function ensureOptionValues() {
```
Comment (colemanw) [inline CRM/Core/BAO/OptionValue.php]: don't forget visibility

Comment (eileenmcnaughton) [inline CRM/Core/BAO/OptionValue.php]: What does visibility even mean?

Comment (colemanw) [inline CRM/Core/BAO/OptionValue.php]: e.g. **public** function ensureOptionValues

Comment (eileenmcnaughton) [inline CRM/Core/BAO/OptionValue.php]: ah - I thought you meant the field in the table

Comment (colemanw) [inline CRM/Core/BAO/OptionValue.php]: Ah, nope don't know what /that/ visibility means :/

**Comment by totten** [comment]
There's a high level question (for which I'm not sure of the answer) about the order/dependency among the upgrade steps. If creating the activity-type has no impact other upgrade-steps (and is unlikely to have such impact in the future), then a declarative style makes sense (and, FWIW, one could piggyback off hook_managed via CRM_Contribute_Info::getManagedEntities). But if there are other changes that need to be coordinated, then I'd stick to hard-coding the incremental-upgrade steps like we've traditionally done. (You can, of course, put conditional in there.)

(As an example: Suppose an upgrade needs to create an activity-type and then migrate some records from an old table to the activity table. Those are dependent upgrade steps. As we accumulate more dependent steps, and it's helpful to have the upgrade-logic be very explicit. But, OTOH, most records have a very boring lifecycle, and it can feel likely overkill.)

*Aside/Personal opinion*: If using the 'incremental upgrade' framework in core, I tend to take the view that one should avoid BAO/DAO/API calls as part of the upgrade logic. That's because BAO/DAO/API logic is subject to change between versions -- which means that the incremental upgrade steps can change in unexpected ways.

**Comment by eileenmcnaughton** [comment]
Very specifically this will have no impact on anything else. More generically I guess you could argue that the 'ensure' function going first would make the order OK.

I guess I'm looking for  a pattern to add option values - because I have seen problems in the past with upgrades falling over because they are trying to create things that are, quite reasonably, already there. And OTOH - bugs on customer sites & in JIRA  because required option values are missing.

I see your point Re BAO/DAO/api calls though.

**Comment by eileenmcnaughton** [comment]
Putting this on the 'ideas to mull over' back burner & have done a PR to add the option value

https://github.com/civicrm/civicrm-core/pull/6062/files

**Pull request closed without merging**